search ui: lucky search toggle menu

limitedmage opened the pull request:
Shows toggle menu for lucky search. Currently, this will only be shown if `defaultPatternType` is `lucky`. Enabling lucky search via this menu will hide the other toggles.

![image](https://user-images.githubusercontent.com/206864/188702800-3ee7dfbc-eef8-49dc-8413-3a9f90d56be1.png)

![image](https://user-images.githubusercontent.com/206864/188702855-4fb74092-c8db-4777-bf06-071c93434c24.png)

## Test plan

- Verify the icon only shows if `defaultPatternType` is `lucky`
- Verify clicking icon opens menu
- Verify clicking on the menu options turns lucky search on and off
- Verify other toggles are only shown if lucky search is off

## App preview:

- [Web](https://sg-web-jp-luckysearchtogglemenu.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-udzkfnujgl.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

almeidapaulooliveira commented:
Some observations:
- Make sure that the close icon is vertically aligned on the header.
- Use the 'Enable' copy unless we want to dynamically change "Enable" and "Enabled" depending on the state.
- The "Suggest variations..." copy appears in bold; can we use the regular style?
- The keyboard shortcut is missing. I'm not sure it's on purpose until we have implemented the keystroke.
- The component's Y position is too high (on top of the icon). Is it possible to align the component with the bottom line of the search field, as in the image below?

<img width="302" alt="Screen Shot 2022-09-06 at 15 20 27" src="https://user-images.githubusercontent.com/64257673/188712132-bae13868-1257-4907-95ca-1d1fdd2cfbf1.png">

rvantonder commented on client/search-ui/src/input/toggles/LuckySearchToggle.tsx:
```diff
@@ -0,0 +1,140 @@
+/**
+ * A toggle displayed in the QueryInput.
+ */
+export const LuckySearchToggle: React.FunctionComponent<LuckySearchToggleProps> = ({
```
Let's rename `Lucky` -> `Smart` everywhere (including file names)? https://sourcegraph.slack.com/archives/C020S8AT3LN/p1662491000273559

I will address backend renaming and exposing activation via URL parameter in follow ups!

rvantonder replied:
I see we're importing some stuff like `import luckyStyles from './LuckySearchToggle.module.scss'`. Feel free to do a rename in a dedicated PR in that case :-)

limitedmage replied:
I'll rename everything before merging :)
When I started working on this last week we hadn't finalized a name yet so I chose to use the existing "lucky" name :)

limitedmage replied:
Stacked PR https://github.com/sourcegraph/sourcegraph/pull/41393 for this.

limitedmage commented:
@almeidapaulooliveira 
Thanks for the feedback! I've addressed all the issues except the keyboard shortcut. Here's how it looks now:

![image](https://user-images.githubusercontent.com/206864/188754486-29d00c29-66a0-4a3a-a0ae-071153d0e623.png)

Because we have not yet decided on what keyboard shortcut to use, I am not doing that work as part of this PR. The keyboard shortcut for this and the other toggles should come later when they are finalized. Please file an issue assigned to me when that is ready.

almeidapaulooliveira commented:
@limitedmage, thanks. It's looking great. 

> Because we have not yet decided what keyboard shortcut to use, I am not doing that work as part of this PR. 

It makes sense; I'll let you know.

rvantonder reviewed (approved):
> Currently, this will only be shown if defaultPatternType is lucky.

Sounds good, this is the right call for now. I checked things locally. We can make it on-by-default along with the changes to remove this value from `defaultPatternType` and expose via URL param. I plan to jump on those changes at the end of next week (so we have 4 weeks of A/B testing on the same UI). I want to keep running the A/B testing after that, but then we can have this UI on by default 🎉. It would be a good idea to log the enable/disable interaction at that time too, I can take care of it.
